### PR TITLE
Get raw block from esplora

### PR DESCRIPTION
**Pull request by lvaccaro**
Get raw block from esplora.
Current esplora endpoint: explorer.bitcoin-lyon.fr/api (no testnet)
Close https://github.com/lvaccaro/esplora_clnd_plugin/issues/3

### CODE REVIEW

**Comment by lvaccaro** [comment]
It looks promising and working good with clightning master.
Need to free data chunks to avoid memory saturation.

File: esplora.c [hunk 1]
```diff
@@ -208,24 +216,23 @@ static struct command_result *getrawblockbyheight(struct command *cmd,
   // Esplora serves raw block
   const char *block_url = tal_fmt(cmd->plugin,
   	"%s/block/%s/raw", endpoint, blockhash);
-	const char *block_res = request_get(block_url);
-	if (!block_res) {
+	struct curl_memory_data *chunk = request_data(block_url, false, NULL);
```
[the remark this answers is not in the document]
Comment (darosior) [reply esplora.c:219]: For more context the pattern we use in Clightning is:
```
char *function_which_allocates(const tal_t *ctx, whatever *A, whatever *B);
```
With ctx the parent to `tal`-allocate from.

**Comment by lvaccaro** [comment]
eheh I tried to use Atom without a lots of success. fixing

**Comment by lvaccaro** [comment]
Think to run clang-format in a next PR, maybe adding it also in travis.

File: esplora.c [hunk 2]
```diff
@@ -208,10 +205,9 @@ static struct command_result *getrawblockbyheight(struct command *cmd,
 	}
 	plugin_log(cmd->plugin, LOG_INFORM, "blockhash: %s from %s", blockhash, blockhash_url);
 
-	// Esplora doesn't serve raw blocks !
-	// Open issue from darosior: https://github.com/Blockstream/esplora/issues/171
-	const char *block_url = tal_fmt(cmd->plugin, 
-		"%s/raw/block/%s", blockchair_endpoint, blockhash);
+  // Esplora serves raw block
+  const char *block_url = tal_fmt(cmd->plugin,
```
[the remark this answers is not in the document]
Comment (lvaccaro) [reply esplora.c]: ta, formatted

File: esplora.c [hunk 3]
```diff
@@ -221,28 +217,15 @@ static struct command_result *getrawblockbyheight(struct command *cmd,
 	}
 
 	// parse rawblock output
-	tokens = json_parse_input(cmd, block_res,
-				  strlen(block_res), &valid);
-	if (!tokens) {
-		err = tal_fmt(cmd, "%s: json error on %s (%.*s)?",
+  const char *rawblock = tal_hex(cmd->plugin, block_res);
+  if (!rawblock) {
```
[the remark this answers is not in the document]
Comment (lvaccaro) [reply esplora.c]: ta, formatted

File: esplora.c [hunk 4]
```diff
@@ -221,28 +217,15 @@ static struct command_result *getrawblockbyheight(struct command *cmd,
...
-	if (!blocktok) {
-		err = tal_fmt(cmd,"%s: had no rawblock for block %s from %s (%.*s)?",
-			      cmd->methodname, blockhash, block_url, (int)sizeof(block_res),
-			      block_res);
-		plugin_log(cmd->plugin, LOG_INFORM, "%s", err);
-		return command_done_err(cmd, BCLI_ERROR, err, NULL);
-	}
-	const char *rawblock = json_strdup(cmd, block_res, blocktok);
-	//plugin_log(cmd->plugin, LOG_INFORM, "rawblock: %s", rawblock);
+  plugin_log(cmd->plugin, LOG_INFORM, "rawblock: %s", rawblock);
```
[the remark this answers is not in the document]
Comment (lvaccaro) [reply esplora.c]: ta, formatted

File: esplora.c [hunk 5]
```diff
@@ -83,10 +83,18 @@ static char *request(const char *url, const bool post, const char* data) {
 	long response_code;
 	curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &response_code);
 	if (response_code != 200) {
+    fprintf(stderr, "response_code != 200\n");
```
[the remark this answers is not in the document]
Comment (lvaccaro) [reply esplora.c]: ta, formatted

File: esplora.c [hunk 6]
```diff
@@ -208,24 +216,23 @@ static struct command_result *getrawblockbyheight(struct command *cmd,
...
-					block_res);
-		plugin_log(cmd->plugin, LOG_INFORM, "%s", err);
-		return command_done_err(cmd, BCLI_ERROR, err, NULL);
-	}
-  plugin_log(cmd->plugin, LOG_INFORM, "rawblock: %s", rawblock);
+    err = tal_fmt(cmd, "%s: convert error on %s",
+					cmd->methodname, block_url);
+    plugin_log(cmd->plugin, LOG_INFORM, "%s", err);
+    return command_done_err(cmd, BCLI_ERROR, err, NULL);
+  }
```
[the remark this answers is not in the document]
Comment (lvaccaro) [reply esplora.c]: ta, formatted

File: esplora.c [hunk 7]
```diff
@@ -208,24 +216,23 @@ static struct command_result *getrawblockbyheight(struct command *cmd,
...
+  //plugin_log(cmd->plugin, LOG_INFORM, "rawblock: %s", rawblock);
```
[the remark this answers is not in the document]
Comment (lvaccaro) [reply esplora.c]: ta, formatted

File: esplora.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (lvaccaro) [reply esplora.c:219]: I could pass ctx to `request_data()`, but there is a realloc inside `write_memory_callback()` implicit called by curl, have you any feedbacks?

[the remark this answers is not in the document]
Comment (darosior) [reply esplora.c:219]: You could tal-allocate the chunk, and add a custom destructor to free all the realocations inside. (A destructor magic callback called by the `tal` subsystem whenever the tal object is freed, thanks Rusty!)

[the remark this answers is not in the document]
Comment (darosior) [reply esplora.c:219]: Could you otherwise realloc as `tal` objects ?

[the remark this answers is not in the document]
Comment (lvaccaro) [reply esplora.c]: `tal_count()` saves me from headache, I like magic.

add tal into b5a57c8d571e0e828c906f0db273fc01964db7dc

[the remark this answers is not in the document]
Comment (darosior) [reply esplora.c]: Tal to replace aspirine in 2020

**Comment by lvaccaro** [comment]
Merged on fast forward

**Comment by lvaccaro** [comment]
Strange error, I think there are no unmerged commits

**Pull request closed without merging**